### ISSUE TITLE
Rule proposal: `isArray()` rather than `instanceof Array`

### ISSUE THREAD

**Issue by iainbeeston** [eslint/eslint#7757]
I'd like to propose a new rule: that checking for `<foo> instanceof Array` and suggesting that it should be rewritten as `Array.isArray(<foo>)`.

Using `instanceof Array` can return false when checking an array object created by another window (or possibly another process on node?), whereas `Array.isArray()` would work as expected. A more in-depth discussion can be found [here](http://web.mit.edu/jwalden/www/isArray.html).

What category of rule is this? (place an "X" next to just one item)
--------

[ ] Enforces code style
[x] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

Provide 2-3 code examples that this rule will warn about:
----

```javascript
if (anArray instanceof Array) {
    // do something...
}
```

```javascript
var foo = anArray instanceof Array ? true : false
```

Why should this rule be included in ESLint (instead of a plugin)?
----

My assumption is that this issue is core to both javascript in the browser and node.js, so I would assume it should go into ESLint itself. `Array.isArray()` has been supported in JavaScript since ES5.

**Comment by platinumazure** [comment]
Thanks for the issue!

My take: I agree with the general merits of why this rule is a good idea. However, due to the sheer number of core ESLint rules, we've had to adopt a stricter standard for rule inclusion in core. That standard is: Would ESLint be considered incomplete without this rule? And I don't think this proposal meets that higher bar.

Others on the team may disagree with me. But if we don't accept this proposal, please do remember that this can absolutely be implemented in a plugin, so that users who want this linting rule can still use it.

**Comment by iainbeeston** [comment]
Thanks for the helpful comments @platinumazure - if that's the consensus I'll look into creating a plugin

**Comment by ljharb** [comment]
@platinumazure what about a more general rule that warns against using `instanceof`, since it doesn't work cross-realm and it's not very reliable anyways? Would that be worth proposing?

**Comment by platinumazure** [comment]
@ljharb You're certainly welcome to lay out a proposal. To be honest, I don't know how I feel about that at this point.

**Comment by mysticatea** [comment]
FWIW, I have the rule in my plugin: https://github.com/mysticatea/eslint-plugin#book-usage

**Comment by sindresorhus** [comment]
You can also find this rule in [eslint-plugin-unicorn](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/no-array-instanceof.md).

**Comment by nzakas** [comment]
Given the existence in plugins already, my preference would be to not add a core rule for this.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.

If you want to use this rule, you can find it in one of multiple plugins linked above, and it's also possible to simulate using selectors with [`no-restricted-syntax`](http://eslint.org/docs/rules/no-restricted-syntax).
